graphql/pupil/mutations.ts

```diff
@@ -361,4 +363,43 @@ export class MutatePupilResolver {
         await invalidatePupilScreening(pupilScreeningId);
         return true;
     }
+
+    @Mutation(() => Boolean)
+    @Authorized(Role.ADMIN)
+    async adminIncreasePupilGrades() {
+        const pupils = await prisma.pupil.findMany({
+            where: {
+                grade: { not: null, notIn: ['13. Klasse', '14. Klasse'] },
+                OR: [{ gradeUpdatedAt: { lt: moment().subtract(10, 'months').toDate() } }, { gradeUpdatedAt: { equals: null } }],
+            },
+        });
+        logger.info(`Attempting to increase the grade of ${pupils.length} pupils`);
+        let validCount = 0;
+        const batchSize = 1000;
+        for (let i = 0; i < pupils.length; i += batchSize) {
+            const batch = pupils.slice(i, i + batchSize);
+
+            await Promise.all(
+                batch.map(async (pupil) => {
+                    try {
+                        const grade = gradeAsInt(pupil.grade);
+                        await prisma.pupil.update({
```

Comment:
Concurrent updates on one table are dangerous and unnecessarily causes lock collisions on the db (in the end it will be sequential) - We had problems in the past were Prisma was giving up on many concurrent transactions and started deadlocking. Can't we use updateMany by grouping by grade? That should be by magnitudes better, as it only requires 13 transactions (one for each grade).

So

for grade in grades:
   get all pupils with grade
   update all pupils with grade
   for pupil in pupils
      take action (Or use the bulk variant?)

Reply:
Alright, I've updated the approach. Thanks for the suggestions!